ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/TransactionTraversalSpec.scala

```diff
@@ -0,0 +1,164 @@
+      case (Raw.NodeId(`rootNid`), _, witnesses) =>
+        witnesses should contain only "Alice"
+        ()
+      case (Raw.NodeId(`fetchNid`), _, witnesses) =>
+        // This is of course ill-authorized, but we check that parent witnesses are included.
+        witnesses should contain.only("Alice", "Bob")
+        ()
+      case what =>
+        fail(s"Traversed to unknown node: $what")
+    }
+  }
+
+  // --------------------------------------------------------
+  // Helpers for constructing transactions.
+
+  case class TransactionBuilder() {
+    private var roots = List.empty[String]
+    private var nodes = Map.empty[String, Node]
+    private var nextNodeId = -1
+
+    def addNode(node: Node.Builder): String = {
+      nextNodeId += 1
+      val nodeId = nextNodeId.toString
+      nodes += nodeId -> node.setNodeId(nodeId).build
+      nodeId
+    }
+
+    def addRoot(node: Node.Builder): String = {
+      // Check that children actually exist.
+      assert(
+        !node.hasExercise ||
+          node.getExercise.getChildrenList.asScala.map(nodes.contains).forall(identity)
+      )
+      val nodeId = addNode(node)
+      roots ::= nodeId
+      nodeId
+    }
+
+    def build: Transaction =
+      Transaction.newBuilder
+        .setVersion(TransactionVersion.minVersion.protoValue)
+        .addAllNodes(nodes.values.asJava)
+        .addAllRoots(roots.reverse.asJava)
+        .build
+  }
+
+  private def withNodeBuilder[A](build: Node.Builder => A): Node.Builder = {
+    val b = Node.newBuilder
+    build(b)
+    b
+  }
+
+  /** Construct a create node. Signatories are the parties whose signature is required to create the
+    * contract. Stakeholders of a contract are the signatories and observers.
+    */
+  private def createNode(signatories: Iterable[String], stakeholders: Iterable[String]) =
+    withNodeBuilder {
+      _.getCreateBuilder
+        .addAllSignatories(signatories.asJava)
+        .addAllStakeholders(stakeholders.asJava)
+    }
+
+  /** Construct a fetch node. Signatories are the signatories of the contract we're fetching.
+    * Stakeholders of a contract are the signatories and observers.
+    */
+  private def fetchNode(signatories: Iterable[String], stakeholders: Iterable[String]) =
+    withNodeBuilder {
+      _.getFetchBuilder
+        .addAllSignatories(signatories.asJava)
+        .addAllStakeholders(stakeholders.asJava)
+    }
+
+  /** Construct an exercise node. Signatories are the signatories of the contract we're exercising
+    * on.
+    */
+  private def exerciseNode(
+      signatories: Iterable[String],
+      stakeholders: Iterable[String],
+      consuming: Boolean,
+      children: String*
+  ) =
+    withNodeBuilder {
+      _.getExerciseBuilder
+        .setConsuming(consuming)
+        .addAllSignatories(signatories.asJava)
+        .addAllStakeholders(stakeholders.asJava)
+        /* NOTE(JM): Actors are no longer included in exercises by the compiler, hence we don't set them */
```

Comment:
I find it hard to believe that a note from someone who hasn't worked on Daml for a year is that useful.